### PR TITLE
RFR: Refactor diff3

### PR DESCRIPTION
**Pull request by aamontal**
Refactor of the core diffing algorithms, including the heckel differ and the three way diffing code. 

Also a pass at adding documentation to many of the customer facing methods.

### CODE REVIEW

File: spec/lib/dyph3/differ/two_way_actions_spec.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+          t2 = ["#<Mido::TextBlock::Id id=\"atom_12225_brief_p_27e015d07debd946ef199ab53ce1ca29_0\">", "#<Mido::TextBlock::Type name=:p>", ":\"Human \"", ":\"activities \"", ":\"probably \"", ":\"caused \"", ":\"the \"", ":\"Holocene \"", ":\"mass \"", ":extinctions", ":\"; \"", ":\"many \"", ":\"methods \"", ":\"have \"", ":\"been \"", ":\"employed \"", ":\"to \"", ":\"estimate \"", ":\"these \"", ":\"extinction \"", ":rates", ":\".\"", "#<Mido::TextBlock::Type closed=true name=:p>", "#<Mido::TextBlock::Id closed=true id=\"atom_12225_brief_p_27e015d07debd946ef199ab53ce1ca29_0\">"]
+          diff = Dyph3::Differ.merge_two_way_diff(t1, t2, diff2: diff2)
```
Comment (mhodgson) [inline spec/lib/dyph3/differ/two_way_actions_spec.rb]: @aamontal @jelder I don’t _really_ care, but do we want to expose MIDO to the world like this?

**Comment by mhodgson** [comment]
Looks really nice!

**Comment by aamontal** [comment]
:blush:

File: spec/lib/dyph3/differ/two_way_actions_spec.rb (hunk 1 above)
Comment (jelder) [inline spec/lib/dyph3/differ/two_way_actions_spec.rb]: This doesn't seem right. It's not even really how Mido works; this is diffing strings that happen to contain pretty-printed objects.

Comment (aamontal) [inline spec/lib/dyph3/differ/two_way_actions_spec.rb]: I can rewrite the the test to check the same expected behavior, with out all the pretty printed stuff

Comment (mhodgson) [inline spec/lib/dyph3/differ/two_way_actions_spec.rb]: :+1:

File: lib/dyph3/differ.rb [hunk 2]
```diff
@@ -1,36 +1,91 @@
 module Dyph3
   class Differ
-    # Algorithm adapted from http://www.rad.upenn.edu/sbia/software/basis/apidoc/v1.2/diff3_8py_source.html
+    #   The differ works on arrays of objects which implement hash and eq.
```
Comment (jelder) [inline lib/dyph3/differ.rb]: Technically, the list is `==`, `eql?`, and `hash`.

**Comment by jelder** [comment]
Awesome.

What it needs now is a `README.md` with examples of how to use it. Please also bring over `Equatable` if you have time.

File: spec/lib/dyph3/differ/misc_conflicts_spec.rb [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        ]}
+
+        it {expect(subject.joined_results).to eq expected_result}
+        it {expect(subject.conflict?).to eq true}
+      end
+
+      describe 'periodic conflicts' do
+        let(:base)  {"alpha\nbase change 1\nbeta\nbase change 2\ndelta\nbase change 3"}
+        let(:left)  {"alpha\nleft change 1\nbeta\nleft change 2\ndelta\nleft change 3"}
+        let(:right) {"alpha\nright change 1\nbeta\nright change 2\ndelta\nright change 3"}
```
Comment (jelder) [inline spec/lib/dyph3/differ/misc_conflicts_spec.rb]: These tests could be a lot easier to read with inline literal arrays like

```
let(:base) { %w[ alpha base_change_1 ] ) # etc
```

**Pull request merged**